### PR TITLE
Removing user typing store

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Removing user typing store from the webapp, and using mattermost-redux for that.

#### Checklist
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [x] Has redux changes (mattermost/mattermost-redux#421)

### CODE REVIEW

**Comment by koxen** [comment]
happy to see another store removed 🥇 !

File: tests/components/create_comment/__snapshots__/create_comment.test.jsx.snap [hunk 1]
```diff
@@ -389,9 +389,9 @@ exports[`components/CreateComment should match snapshot, empty comment 1`] = `
         </span>
       </div>
     </div>
-    <MsgTyping
+    <Connect(MsgTyping)
```
Comment (koxen) [inline tests/components/create_comment/__snapshots__/create_comment.test.jsx.snap]: ? why do you need a connect here, you already connceted it in msg_typing/index.js

File: actions/websocket_actions.jsx [hunk 2]
```diff
@@ -524,10 +525,36 @@ function handlePreferencesDeletedEvent(msg) {
 }
 
 function handleUserTypingEvent(msg) {
-    GlobalActions.emitRemoteUserTypingEvent(msg.broadcast.channel_id, msg.data.user_id, msg.data.parent_id);
+    const state = getState();
```
Comment (koxen) [inline actions/websocket_actions.jsx]: why are you duplicating code with mattermost-redux on this? https://github.com/mattermost/mattermost-redux/blob/master/src/actions/websocket.js#L620

Comment (koxen) [reply actions/websocket_actions.jsx]: ok so I see the websocket client can only take a one event handler callback...         websocketClient.setEventCallback(handleEvent);
we should change it to allow more than 1 callback and only leave webapp-specific code inside websocket_actions(that redux is not taking care of); really don't like the way it is currently duplicated

File: tests/components/create_comment/__snapshots__/create_comment.test.jsx.snap (hunk 1 above)
Comment (koxen) [reply tests/components/create_comment/__snapshots__/create_comment.test.jsx.snap]: sorry, didn't realize I was looking at a snapshot file. ignore the comment

File: components/msg_typing/msg_typing.jsx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            />
+        );
+    }
+
+    render() {
+        return (
+            <span className='msg-typing'>{this.getTypingText()}</span>
+        );
+    }
+}
```
[the remark this answers is not in the document]
Comment (jespino) [reply components/msg_typing/msg_typing.jsx]: Yeah, you are right :), I going to add a component test for this :)

File: actions/websocket_actions.jsx (hunk 2 above)
Comment (jespino) [reply actions/websocket_actions.jsx]: Us neither, the websocket client code is duplicated in the mattermost-redux and mattermost-webapp, we have to find the time to migrate it to use only the mattermost-redux and remove entirely the websocket client from the mattermost-webapp. Until them, I had to replicate the behavior in the mattermost-webapp repo.

File: components/msg_typing/msg_typing.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (jespino) [reply components/msg_typing/msg_typing.jsx]: Done! :)

File: stores/post_store.jsx [hunk 4]
```diff
@@ -202,6 +203,7 @@ PostStore.dispatchToken = AppDispatcher.register((payload) => {
         dispatch({...action, type: ActionTypes.EDIT_POST});
         break;
     case ActionTypes.RECEIVED_POST_SELECTED:
+        dispatch({data: action.postId, type: PostTypes.RECEIVED_POST_SELECTED});
```
Comment (lieut-data) [inline stores/post_store.jsx]: I'm not super familiar with our Flux->Redux compatibility layer. Can you elaborate on this addition?

File: actions/global_actions.jsx [hunk 5]
```diff
@@ -405,23 +406,24 @@ export function loadDefaultLocale() {
...
-    const license = getLicense(getState());
-    const config = getConfig(getState());
-    if (license.IsLicensed === 'true' && config.ExperimentalTownSquareIsReadOnly === 'true') {
-        const channel = ChannelStore.getChannelById(channelId);
-        if (channel && ChannelStore.isDefault(channel)) {
-            return;
+export function emitLocalUserTypingEvent(channelId, parentPostId) {
+    const userTyping = async (actionDispatch, actionGetState) => {
+        const state = actionGetState();
+        const config = state.entities.general.config;
```
Comment (lieut-data) [inline actions/global_actions.jsx]: Can you use the `getConfig` selector exported from Redux?

File: actions/global_actions.jsx [hunk 6]
```diff
@@ -405,23 +406,24 @@ export function loadDefaultLocale() {
...
-    const membersInChannel = ChannelStore.getStats(channelId).member_count;
-
-    const license = getLicense(getState());
-    const config = getConfig(getState());
-    if (license.IsLicensed === 'true' && config.ExperimentalTownSquareIsReadOnly === 'true') {
-        const channel = ChannelStore.getChannelById(channelId);
-        if (channel && ChannelStore.isDefault(channel)) {
-            return;
+export function emitLocalUserTypingEvent(channelId, parentPostId) {
+    const userTyping = async (actionDispatch, actionGetState) => {
```
Comment (lieut-data) [inline actions/global_actions.jsx]: nit: why are these prefixed with `action`? Is there something action-specific about this version of `dispatch` and `getState`?

File: actions/websocket_actions.jsx [hunk 7]
```diff
@@ -524,10 +525,36 @@ function handlePreferencesDeletedEvent(msg) {
...
+    const {currentUserId, profiles, statuses} = state.entities.users;
```
Comment (lieut-data) [inline actions/websocket_actions.jsx]: Worth considering `getCurrentUserId` and the like for these?

File: stores/post_store.jsx (hunk 4 above)
Comment (jespino) [reply stores/post_store.jsx]: Yes. `ActionTypes.RECIEVED_POST_SELECTED` is managed by the AppDispatcher, and is an "Old-flux" event which is not managed by redux at all. In the redux we have a reducer to manage `PostTypes.RECEIVED_POST_SELECTED` to update the `selectedPost` in the global store, that information was used by the `getUsersTyping` selector, but in not generated by the webapp (before this change). Finally I'm using my own selector which don't use the selectedPost information in the webapp, but anyway, makes sense to dispatch that event to maintain the consistency. I think.

File: actions/global_actions.jsx (hunk 5 above)
Comment (jespino) [reply actions/global_actions.jsx]: Yes, makes sense :)

File: actions/global_actions.jsx (hunk 6 above)
Comment (jespino) [reply actions/global_actions.jsx]: Well, basically is to no "shade" the global dispatch and getState variables, avoiding this way the es-lint warning.

File: actions/websocket_actions.jsx (hunk 7 above)
Comment (jespino) [reply actions/websocket_actions.jsx]: Yes, I can try to use selectors there too. I'll do the changes.

File: actions/global_actions.jsx (hunk 6 above)
Comment (lieut-data) [reply actions/global_actions.jsx]: Ah, that makes sense given the transitional nature of this file.

File: actions/global_actions.jsx (hunk 5 above)
Comment (koxen) [reply actions/global_actions.jsx]: I'd say it's not worth fixing this _here_ since it's a copypaste from mattermost-redux...

Comment (lieut-data) [reply actions/global_actions.jsx]: Happy to defer to @jespino's best judgement here.

File: actions/websocket_actions.jsx (hunk 2 above)
Comment (jwilander) [reply actions/websocket_actions.jsx]: The ticket to work on that is here if anyone wants to give it a shot :) It's something I've wanted to get done for a while, just haven't had the time
https://mattermost.atlassian.net/browse/MM-8648

File: actions/global_actions.jsx (hunk 5 above)
Comment (jespino) [reply actions/global_actions.jsx]: Done, wasn't an effort at all :)

File: actions/websocket_actions.jsx (hunk 7 above)
Comment (jespino) [reply actions/websocket_actions.jsx]: Done! :)

**Pull request merged**